### ISSUE TITLE
examples/mnli_example:  run_mnli_train.sh 模型没被训练

### ISSUE THREAD

**Issue by Yin169** [airaria/TextBrewer#67]
你好，

目前正在尝试了解并且跑通mnli_example， 但当跑 run_mnli_train.sh 脚本时发现模型 acc 在训练过程中并无变化，并且保持很低的一个分数。

**Comment by airaria** [comment]
请描述详细些，比如PyTorch和Transformer版本是多少？
对sh脚本和代码有没有做哪些调整？

**Comment by Yin169** [comment]
torch==1.5.1+cu101,  transformers==3.0.2. 对代码和脚本没有调整

**Comment by qi2qi** [comment]
你好，想问下train MNLI teacher model的时候参数就是lr=2 * e-5 然后3个epoch就可以达到80+ %的acc么？
一开始用预设的参数我也有acc很低的问题（～35%），稍微调大了lr 然后改为5个epoch换了个torch seed之后现在acc可以达到66%左右。不知道是哪里不太对么？
我用的readme里推荐的版本torch 1.2.0 transformer 3.0.2，code里面只是把load checkpoint那里改成直接从transformer load pretrain bert uncase。

**Comment by airaria** [comment]
@Yin169 @qi2qi 可以把日志贴出来吗，在train开始之前的部分？

**Comment by qi2qi** [comment]
@airaria hello这是开始的log部分
2021/05/24 22:46:22 - INFO - transformers.file_utils -  PyTorch version 1.2.0 available.
2021/05/24 22:46:23 - INFO - transformers.file_utils -  TensorFlow version 2.2.0 available.
2021/05/24 22:46:24 - INFO - Main -  output_dir:../../output_root/mnli_base_lr3e15_bs32_teacher
2021/05/24 22:46:24 - INFO - Main -  data_dir:../../data/MNLI
2021/05/24 22:46:24 - INFO - Main -  max_seq_length:128
2021/05/24 22:46:24 - INFO - Main -  do_train:True
2021/05/24 22:46:24 - INFO - Main -  do_predict:True
2021/05/24 22:46:24 - INFO - Main -  train_batch_size:32
2021/05/24 22:46:24 - INFO - Main -  predict_batch_size:8
2021/05/24 22:46:24 - INFO - Main -  learning_rate:3e-05
2021/05/24 22:46:24 - INFO - Main -  num_train_epochs:15.0
2021/05/24 22:46:24 - INFO - Main -  warmup_proportion:0.1
2021/05/24 22:46:24 - INFO - Main -  no_cuda:False
2021/05/24 22:46:24 - INFO - Main -  gradient_accumulation_steps:1
2021/05/24 22:46:24 - INFO - Main -  local_rank:-1
2021/05/24 22:46:24 - INFO - Main -  fp16:True
2021/05/24 22:46:24 - INFO - Main -  random_seed:9577
2021/05/24 22:46:24 - INFO - Main -  weight_decay_rate:0.01
2021/05/24 22:46:24 - INFO - Main -  do_eval:True
2021/05/24 22:46:24 - INFO - Main -  PRINT_EVERY:200
2021/05/24 22:46:24 - INFO - Main -  ckpt_frequency:2
2021/05/24 22:46:24 - INFO - Main -  temperature:1
2021/05/24 22:46:24 - INFO - Main -  teacher_cached:False
2021/05/24 22:46:24 - INFO - Main -  task_name:mnli
2021/05/24 22:46:24 - INFO - Main -  aux_task_name:None
2021/05/24 22:46:24 - INFO - Main -  aux_data_dir:None
2021/05/24 22:46:24 - INFO - Main -  matches:None
2021/05/24 22:46:24 - INFO - Main -  model_config_json:../../output_root/mnli_base_lr3e15_bs32_teacher/TrainBertTeacher.json.run
2021/05/24 22:46:24 - INFO - Main -  do_test:False
2021/05/24 22:46:24 - WARNING - Main -  Output directory () already exists and is not empty.
2021/05/24 22:46:24 - INFO - Main -  device cuda n_gpu 2 distributed training False
2021/05/24 22:46:24 - INFO - utils -  Loading features from cached file ../../data/MNLI/bertuncased_train_128_mnli
2021/05/24 22:46:37 - INFO - utils -  Loading features from cached file ../../data/MNLI/bertuncased_dev_128_mnli
2021/05/24 22:46:37 - INFO - utils -  Loading features from cached file ../../data/MNLI/bertuncased_dev_128_mnli-mm
2021/05/24 22:46:38 - INFO - Main -  Data loaded
2021/05/24 22:46:39 - INFO - transformers.configuration_utils -  loading configuration file https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-uncased-config.json from cache at ../.cache/torch/transformers/4dad0251492946e18ac39290fcfe91b89d370fee250efe9521476438fe8ca185.7156163d5fdc189c3016baca0775ffce230789d7fa2a42ef516483e4ca884517
2021/05/24 22:46:39 - INFO - transformers.configuration_utils -  Model config BertConfig {
  "architectures": [
    "BertForMaskedLM"
  ],
  "attention_probs_dropout_prob": 0.1,
  "gradient_checkpointing": false,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "layer_norm_eps": 1e-12, 
  "max_position_embeddings": 512,
  "model_type": "bert",
  "num_attention_heads": 12,
  "num_hidden_layers": 12, 
  "output_hidden_states": true,
  "pad_token_id": 0,
  "type_vocab_size": 2,
  "vocab_size": 30522
}

2021/05/24 22:46:39 - INFO - transformers.modeling_utils -  loading weights file https://cdn.huggingface.co/bert-base-uncased-pytorch_model.bin from cache at /.cache/torch/transformers/f2ee78bdd635b758cc0a12352586868bef80e47401abe4c4fcc3832421e7338b.36ca03ab34a1a5d5fa7bc3d03d55c4fa650fed07220e2eeebc06ce58d0e9a157 

2021/05/24 22:46:40 - INFO - transformers.modeling_utils -  All model checkpoint weights were used when initializing BertModel.

2021/05/24 22:46:40 - INFO - transformers.modeling_utils -  All the weights of BertModel were initialized from the model checkpoint at bert-base-uncased.
If your task is similar to the task the model of the ckeckpoint was trained on, you can already use BertModel for predictions without further training.
2021/05/24 22:46:45 - INFO - Main -  Length of all_trainable_params: 2
check data parallel True   
2021/05/24 22:46:45 - INFO - Main -  ***** Running training *****
2021/05/24 22:46:45 - INFO - Main -    Num examples = 392702
2021/05/24 22:46:45 - INFO - Main -    Forward batch size = 32
2021/05/24 22:46:45 - INFO - Main -    Num backward steps = 184065
Selected optimization level O1:  Insert automatic casts around Pytorch functions and Tensor methods.

Defaults for this optimization level are:
enabled                : True
opt_level              : O1
cast_model_type        : None
patch_torch_functions  : True
keep_batchnorm_fp32    : None
master_weights         : None
loss_scale             : dynamic
Processing user overrides (additional kwargs that are not None)...
After processing overrides, optimization options are:
enabled                : True
opt_level              : O1
cast_model_type        : None
patch_torch_functions  : True
keep_batchnorm_fp32    : None
master_weights         : None
loss_scale             : dynamic
2021/05/24 22:46:45 - INFO - Distillation -  Training steps per epoch: 12271
2021/05/24 22:46:45 - INFO - Distillation -  Checkpoints(step): [0, 6135]
  0%|                                                                                                                                                                 | 0/15 [00:00<?, ?it/s]2021/05/24 22:46:45 - INFO - Distillation -  Epoch 1
2021/05/24 22:46:45 - INFO - Distillation -  Length of current epoch in forward batch: 12271
                  /ldap_home/kiki.wang/my_env/lib/python3.6/site-packages/torch/nn/parallel/_functions.py:61: UserWarning: Was asked to gather along dimension 0, but all input tensors were scalars; will instead unsqueeze and return a vector.
  warnings.warn('Was asked to gather along dimension 0, but all '
/ldap_home/kiki.wang/my_env/lib/python3.6/site-packages/torch/optim/lr_scheduler.py:73: UserWarning: Seems like `optimizer.step()` has been overridden after learning rate scheduler initialization. Please, make sure to call `optimizer.step()` before `lr_scheduler.step()`. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
  "https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate", UserWarning)
                        2021/05/24 22:52:42 - INFO - Distillation -  Global step: 613, epoch step:613
612it [05:57,  1.74it/s] 2021/05/24 22:58:36 - INFO - Distillation -  Global step: 1226, epoch step:1226
1225it [11:50,  1.74it/s 2021/05/24 23:04:30 - INFO - Distillation -  Global step: 1839, epoch step:1839
1838it [17:44,  1.74it/s]2021/05/24 23:10:24 - INFO - Distillation -  Global step: 2452, epoch step:2452
2451it [23:38,  1.74it/s]2021/05/24 23:16:18 - INFO - Distillation -  Global step: 3065, epoch step:3065
3064it [29:33,  1.73it/s]2021/05/24 23:22:12 - INFO - Distillation -  Global step: 3678, epoch step:3678
3677it [35:27,  1.74it/s]2021/05/24 23:28:06 - INFO - Distillation -  Global step: 4291, epoch step:4291
4290it [41:21,  1.74it/s]Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 131072.0
4460it [42:59,  1.73it/s]2021/05/24 23:34:00 - INFO - Distillation -  Global step: 4904, epoch step:4904
4903it [47:15,  1.73it/s]  
5468it [52:41,  1.74it/s]  
5482it [1:33:03,  7.59s/it]2021/05/25 00:20:08 - INFO - Distillation -  Global step: 5517, epoch step:5517
5516it [1:33:22,  1.72it/s]2021/05/25 00:26:02 - INFO - Distillation -  Global step: 6130, epoch step:6130
6129it [1:39:17,  1.73it/s]2021/05/25 00:26:05 - INFO - Distillation -  Saving at global step 6135, epoch step 6135 epoch 1
2021/05/25 00:26:06 - INFO - Distillation -  Running callback function...
2021/05/25 00:26:06 - INFO - predict_function -  Predicting...
2021/05/25 00:26:06 - INFO - predict_function -  ***** Running predictions *****
2021/05/25 00:26:06 - INFO - predict_function -   task name = mnli
2021/05/25 00:26:06 - INFO - predict_function -    Num  examples = 9815
2021/05/25 00:26:06 - INFO - predict_function -    Batch size = 8
Evaluating: 100%|########################################################################################################################################| 1227/1227 [01:57<00:00, 10.47it/s]
2021/05/25 00:28:03 - INFO - predict_function -  ***** Eval results 6135 task mnli *****
2021/05/25 00:28:03 - INFO - predict_function -  mnli acc = 0.50168#####################################################################################8| 1226/1227 [01:57<00:00, 10.49it/s]
2021/05/25 00:28:03 - INFO - predict_function -  Predicting...
2021/05/25 00:28:03 - INFO - predict_function -  ***** Running predictions *****
2021/05/25 00:28:03 - INFO - predict_function -   task name = mnli-mm
2021/05/25 00:28:03 - INFO - predict_function -    Num  examples = 9832
2021/05/25 00:28:03 - INFO - predict_function -    Batch size = 8
Evaluating: 100%|########################################################################################################################################| 1229/1229 [01:57<00:00, 10.47it/s]
2021/05/25 00:30:00 - INFO - predict_function -  ***** Eval results 6135 task mnli-mm *****
2021/05/25 00:30:00 - INFO - predict_function -  mnli-mm acc = 0.50844##################################################################################8| 1228/1229 [01:57<00:00, 10.61it/s]

**Comment by airaria** [comment]
我查一下示例有没有问题。
你们可以checkout到fb009a6，先用那里的MNLI example

**Comment by airaria** [comment]
例子我又运行了一遍，第一个checkpoint的效果可以达到80以上

**Comment by qi2qi** [comment]
@airaria 好的谢谢，所以确认一下你student json里面的checkpoint放的是bert uncase pretrain么？vocab也是30522的那个bert的vocab.txt么？lr=2e-5?
其他的参数：
accu=1
temperature=8
batch_size=32
length=128
都是这些对么？
如果是的话，那我再去试一下。

**Comment by qi2qi** [comment]
@airaria hello ziqing, 麻烦问一下distill一个epoch的MNLI task需要多久么？用多少的gpu呢？
我现在想用这个distill一个一千万数据的bert model，2个gpu，一个epoch需要20小时，并且会有Gradient overflow 的问题，会不会是哪里不对呢？
  2%|###7                                                                                                                                                | 1/40 [21:09:55<825:26:50, 76195.15s/it]2021/05/31 08:43:14 - INFO - Distillation -  Epoch 2
2021/05/31 08:43:14 - INFO - Distillation -  Length of current epoch in forward batch: 156250
                         Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 8388608.0
1479it [12:02,  2.06it/s]Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 8388608.0
3529it [28:40,  2.06it/s]Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 8388608.0
5551it [45:06,  2.08it/s]  Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 8388608.0
7569it [1:01:33,  2.06it/s]2021/05/31 09:46:42 - INFO - Distillation -  Global step: 164052, epoch step:7802
7801it [1:03:26,  2.06it/s]Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 8388608.0
9574it [1:17:52,  2.06it/s] Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 8388608.0
11604it [1:34:25,  2.01it/s]Gradient overflow.  Skipping step, loss scaler 0 reducing loss scale to 8388608.0

**Comment by airaria** [comment]
@qi2qi 对，用的是BERT-uncased，在MNLI上其他的超参为
max_seq_length 128
train_batch_size 32
random_seed 9580
num_train_epochs 3
learning_rate 2e-5
gradient_accumulation-steps 1
使用一张V100 32GB训练，3个epoch大约需要1.5个小时


在训练刚开始时出现Gradient overflow是正常的

**Comment by Jay2Coomzz** [comment]
请问MNLI数据集哪里有提供呢？我没找到，只找到了一个MultiNLI。

**Comment by airaria** [comment]
Use MultiNLI matched/mismatched from GLUE benchmark:

https://gluebenchmark.com/tasks







在 2021-06-10 15:18:06，"Jay2Coomzz" ***@***.***> 写道：

请问MNLI数据集哪里有提供呢？我没找到，只找到了一个MultiNLI。
